### ISSUE TITLE
`unref` support for dns requests

### ISSUE THREAD

**Issue by silverwind** [libuv/libuv#203]
Under certain network conditions, dns lookups can run for up to 30s (the glibc timeout) without returning any error, and in turn preventing the main io.js/node thread from exiting. Support for `unref`'ed dns lookup would solve the exiting issue. See:

https://github.com/joyent/node/issues/7149
https://github.com/iojs/io.js/issues/795

**Comment by bnoordhuis** [comment]
For completeness' sake, the thread is blocked in getaddrinfo() at that point and not easily interruptible.  I think workers would have to maintain a flag that's updated for each work request that tells the main thread if the worker needs joining on shutdown.

**Comment by saghul** [comment]
I don't think this problem can be solved in 100% of the cases, but we can try :-) If the user runs a CPU intensive function via `uv_queue_work` there is very little we can do. Maybe we can try to use pthread_cancel as part of `uv_cancel`, for certain types of requests.

Now, that said, I think a good way forward is to traeat the threadpool as a handle where some operations run. That handle can be unref'd, and if that is the case the application would exit without waiting for the threads to finish. I outlined some ideas about that here: https://github.com/libuv/leps/pull/4 even though we haven't made any progress in any direction yet.

Basically Node could use multiple threadpools, one for DNS, another one for files, etc. This way we put the burden of maintenance to the application, outside of libuv :-)

**Comment by ogapo** [comment]
Would it make more sense to implement DNS resolution (on linux at least) using getaddrinfo_a (http://man7.org/linux/man-pages/man3/getaddrinfo_a.3.html) so it can be properly cancelled rather than trying to club the whole thread?

**Comment by saghul** [comment]
@ogapo Not really. AFAIK getaddrinfo_a is implemented using threads anyway. Our implementation runs on the threadpool and is already cancelable (if it's not running yet). What is currently not implemented is making this running requests unref-able so the event loop can end early.

**Comment by bminer** [comment]
I have a question.  Even if certain threads could be cancelled using `pthread_cancel` as part of `uv_cancel`, how would you know when to cancel them?  For example, if a user calls `process.exit(0)` from a Node process, how does that signal propagate down to libuv land?

**Comment by bminer** [comment]
@saghul ??
@bnoordhuis ??

Anybody.... (echo) anybody.... ?  :)

I'm wondering about the best way to cancel threads in libuv.  It might be a rather important topic in the future as developers look for a way to cancel pending Promises or other long-running operations...

Not to mention... that `process.exit(0)` currently causes Node to hang until all threads are completed.  :-O  This just *seems* wrong, especially when things like DNS resolution can take quite a while to fail.

**Comment by bminer** [comment]
See https://github.com/node-hid/node-hid/issues/194 for another example of this...

**Comment by saghul** [comment]
I'm afraid I don't have a good answer for this.We might need to add unref-able requests, but that won't help with thread cancellation.

Nobody is working on this at the moment, AFAICT.

**Comment by bminer** [comment]
I wouldn't mind working on this, but I think it would be worthwhile to discuss a bit more about it and reach a consensus.  When Node is given a signal to exit (i.e. `process.exit` is called), perhaps libuv threads should be cancelled?  How would this work?

Is there any other use case for cancelling a thread?  For example, if a library is designed to spawn multiple threads to do CPU-heavy work, what mechanism is there to cancel those threads?

The [libuv book](https://nikhilm.github.io/uvbook/threads.html#id1) says: "A well designed program would have a way to terminate long running workers that have already started executing. Such a worker could periodically check for a variable that only the main process sets to signal termination."

Are we leaving it to library authors to add hooks to [Node's "exit" event](https://nodejs.org/api/process.html#process_event_exit) and then either close underlying resources (i.e. close a file descriptor, which ultimately causes a blocking read to fail and the thread to terminate) or set a flag in the main process to signal termination to the running threads?  Or, should there be a mechanism to treat `process.exit` as a special case and terminate as many threads as possible with `pthread_cancel`?

Thoughts?

**Comment by Fishrock123** [comment]
So me and @trevnorris chatted about this problem recently and our conclusion was basically "throw dns requests in a separate threadpool and just kill the thread when you want stuff to end", or close to that, anyway.

**Comment by addaleax** [comment]
If that’s a route to go down: /cc @davisjam

**Comment by bnoordhuis** [comment]
Me and @trevnorris talked about that some years ago, IIRC, but the problem with `pthread_cancel()` is that it's buggy in many libcs. Hardly anyone uses thread cancellation because it's so fraught with dangers and hence bugs in the implementation don't get shook out.

**Comment by Fishrock123** [comment]
So, as a continuation, I talked a bit with @davisjam yesterday and we concluded that just abandoning the thread might actually be the safer option, since killing `getaddrinfo` can cause state corruption...

There's also some details around https://github.com/libuv/libuv/pull/1726 which should make all of this easier (I think).

**Comment by davisjam** [comment]
Some details on the (non-)cancellability of `getaddrinfo` and `getnameinfo` for those interested:

1. [POSIX safety concepts](https://www.gnu.org/software/libc/manual/html_node/POSIX-Safety-Concepts.html) defines `AC-Safe` for "safe to cancel this", and remarks that "Functions not explicitly documented as safe in a safety context should be regarded as Unsafe."
2. [getaddrinfo man page](http://man7.org/linux/man-pages/man3/getaddrinfo.3.html) says it is `MT-Safe` (no mention made of `AC-safe`ty).
3. [getnameinfo man page](http://man7.org/linux/man-pages/man3/getnameinfo.3.html) says it is `MT-Safe` (no mention made of `AC-safe`ty).

And yes, per discussion I agree that #1726 would make fancier strategies easier for Node. But perhaps the default libuv TP should do something smart to benefit other embedders, in the spirit of #1845? (Though AFAIK no one has complained DNS issues except for Node.js).

**Comment by ogapo** [comment]
Avoiding stale closure. I still think this would be nice. Sounds like this is mostly an at exit issue, so abandoning the thread if we know the dns request is unref'd (and tracking that as necessary) sounds like a good approach to me.

**Comment by saghul** [comment]
If nobody is willing to do the work (and by the looks of it there isn't), I don't see the point on keeping this open. The issue will remain the GH repo for future reference, should anyone want to pick it up.

**Comment by bnoordhuis** [comment]
I've tagged this 'known limitation' for now and I'll go ahead and close it out. I'm all ears if someone has a brilliant idea.
